### PR TITLE
[Gas Estimation] Implement block-based gas estimation

### PR DESCRIPTION
**Pull request by bchocho**
### Description

Implementation of [AIP-34](https://github.com/aptos-foundation/AIPs/blob/main/aips/aip-34.md)

### Test Plan

Added api tests and smoke test.

### CODE REVIEW

File: api/src/context.rs
Comment (JoshLind) [inline api/src/context.rs:849]: nit: do we want this to be configurable, too?

File: api/goldens/aptos_api__tests__transactions_test__test_gas_estimation_one_block.json
Comment (JoshLind) [inline api/goldens/aptos_api__tests__transactions_test__test_gas_estimation_one_block.json:2]: FWIW, these labels aren't as descriptive to me as those used in the AIP. Should we consider changing them? For example: `low gas fee`, `market gas fee`, `aggressive (or priority) gas fee` seem more explanatory? 😄

Also, silly question, do we define these labels anywhere? The spec doesn't seem super descriptive: https://fullnode.devnet.aptoslabs.com/v1/spec#/operations/estimate_gas_price. Might be worth updating the spec and/or docs to call the new bands out?

File: api/src/context.rs
Comment (JoshLind) [inline api/src/context.rs:962]: Do we want to add a unit test to ensure we are doing garbage collection, i.e., that the cache never grows indefinitely?

File: api/src/context.rs
Comment (JoshLind) [inline api/src/context.rs:902]: nit: should this code block be in a helper function?

**Review by JoshLind — COMMENTED** [review]
Awesome, thanks @bchocho 😄 First pass looks reasonable to me -- only nits and questions.

File: api/goldens/aptos_api__tests__transactions_test__test_gas_estimation_one_block.json
Comment (bchocho) [reply api/goldens/aptos_api__tests__transactions_test__test_gas_estimation_one_block.json:2]: (1) Good callout about changing the spec/docs. I'll update the PR with these changes.

(2) I'm not sure what breaks if we change the json field names, so I'm playing it safe by reusing what's already there.
@gregnazario do you have any insight here? Is it possible to change the names, or would we have to create a new API endpoint and then deprecate this one?

File: api/src/context.rs
Comment (bchocho) [reply api/src/context.rs:914]: I guess it's a style issue - I'd rather not put this into a helper function as it has a few shared variables with the other parts and won't be reusable.

**Comment by bchocho** [comment]
@JoshLind thanks! I've addressed your comments (except replied to the last one)

File: api/src/context.rs
Comment (JoshLind) [reply api/src/context.rs]: Makes sense 😄 Generally, I'll try add configs for things that I suspect will be worth tuning at runtime, so that I have some levers to pull in case things don't behave as expected, but I get that there's a trade-off. Your call 😃

**Review by JoshLind — APPROVED** [review]
Nice, thanks @bchocho! Looks solid to me 😄

**Comment by bchocho** [comment]
@JoshLind I actually updated the spec a bit, as reflected in https://github.com/aptos-foundation/AIPs/pull/145. Hope this makes sense!

File: api/src/context.rs
Comment (bchocho) [reply api/src/context.rs:966]: Generally I agree. In these cases though we are using a fraction of the vector length, so I'm highly confident there's no out of bounds error that we can hit.

Comment (JoshLind) [reply api/src/context.rs:966]: > so I'm highly confident there's no out of bounds error that we can hit.

Bold 😄 SGTM.

Comment (bchocho) [reply api/src/context.rs]: I just took your advice and used get. Not feeling so bold anymore :P

**Review by pr-aptos — APPROVED** [review]
stamping to unblock devnet

**Pull request merged**
